### PR TITLE
[RFC] tests: Migrate legacy test 80.

### PR DESCRIPTION
**Pull request by lucc**
The test is also split in several blocks and heavily modernized.  This was
done to prevent the following quoting and escaping problems during migration:
- the vim command `put =...` treats double quotes as the start of a comment so
  they have to be escaped with a backslash
- when inserting control characters on the command line they have to be
  escaped with <C-V>

### CODE REVIEW

File: test/functional/legacy/080_substitute_spec.lua [hunk 1]
```diff
@@ -0,0 +1,181 @@
+-- Test for *sub-replace-special* and *sub-replace-expression* on substitue().
+-- Test for submatch() on substitue().
+-- Test for *:s%* on :substitute.
+
+local helpers = require('test.functional.helpers')
+local feed, insert, source = helpers.feed, helpers.insert, helpers.source
+local clear, execute, expect = helpers.clear, helpers.execute, helpers.expect
+local eq, eval = helpers.eq, helpers.eval
+
+local lastline = function(text)
```
Comment (Pyrohh) [inline test/functional/legacy/080_substitute_spec.lua]: I might be missing something, but I don't see this function used anywhere.

Comment (lucc) [inline test/functional/legacy/080_substitute_spec.lua]: you are right, I forgot to delete it after debugging. will do it now.

File: test/functional/legacy/080_substitute_spec.lua [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    eq('a1a2a3a', eval([[substitute('123', '\ze', 'a', 'g')]]))
+    eq('a1a2a3', eval([[substitute('123', '\ze.', 'a', 'g')]]))
+    eq('aaa', eval([[substitute('123', '.\ze', 'a', 'g')]]))
+    eq('aa2a3a', eval([[substitute('123', '1\|\ze', 'a', 'g')]]))
+    eq('1aaa', eval([[substitute('123', '1\zs\|[23]', 'a', 'g')]]))
+  end)
+end)
+
+describe(':substitue', function()
+  before_each(clear)
```
Comment (Pyrohh) [inline test/functional/legacy/080_substitute_spec.lua]: Do you know of any tests which don't require this? I wonder if it'd be easily doable to have this always done.

Comment (lucc) [inline test/functional/legacy/080_substitute_spec.lua]: There might be tests that require *more than this*. It is described in some comment on the issues that describe writing and conversion of tests here on GH. I don't know where exactly.

Comment (Pyrohh) [inline test/functional/legacy/080_substitute_spec.lua]: Okay, I suppose it's fine as is (for now).

File: test/functional/legacy/080_substitute_spec.lua [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    eq('t\rt', eval([[substitute('tTt', 'T', "\r", '')]]))
+    eq('u\nu', eval([[substitute('uUu', 'U', "\n", '')]]))
+    eq('v\bv', eval([[substitute('vVv', 'V', "\b", '')]]))
+    eq('w\\w', eval([[substitute('wWw', 'W', "\\", '')]]))
+    -- See the line with t's above.
+    eq('YyyY', eval([[substitute('Y', 'Y', '\L\uyYy\l\EY', '')]]))
+    eq('zZZz', eval([[substitute('Z', 'Z', '\U\lZzZ\u\Ez', '')]]))
+  end
+
+  it('with "set magic" (TEST_1)', function()
```
Comment (Pyrohh) [inline test/functional/legacy/080_substitute_spec.lua]: What's the reason for having the `TEST_X` annotation? I haven't seen it done elsewhere... perhaps it's to ease reviews?

File: test/functional/legacy/080_substitute_spec.lua [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    eq('m\rm', eval([[substitute('mMm', 'M', '\r', '')]]))
+    eq('n\rn', eval("substitute('nNn', 'N', '\\\r', '')"))
+    eq('o\no', eval([[substitute('oOo', 'O', '\n', '')]]))
+    eq('p\bp', eval([[substitute('pPp', 'P', '\b', '')]]))
+    eq('q\tq', eval([[substitute('qQq', 'Q', '\t', '')]]))
+    eq('r\\r', eval([[substitute('rRr', 'R', '\\', '')]]))
+    eq('scs', eval([[substitute('sSs', 'S', '\c', '')]]))
+    -- There is no line with t's in the original legacy test part 1.  The same
+    -- test was done with x's.  Part two did the following test with t's and
+    -- moved the test here done with y and z up to x and y.
```
Comment (Pyrohh) [inline test/functional/legacy/080_substitute_spec.lua]: Perhaps this should comment should be in the commit message instead; I don't think it would be useful to a future reader of the test.

File: test/functional/legacy/080_substitute_spec.lua [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+local helpers = require('test.functional.helpers')
+local feed, insert, source = helpers.feed, helpers.insert, helpers.source
+local clear, execute, expect = helpers.clear, helpers.execute, helpers.expect
+local eq, eval = helpers.eq, helpers.eval
+
+describe('substitue()', function()
+  before_each(clear)
+
+  local function test_1_and_2()
```
Comment (Pyrohh) [inline test/functional/legacy/080_substitute_spec.lua]: Given this is called below, I think the name should be more descriptive.

File: test/functional/legacy/080_substitute_spec.lua (hunk 3 above)
Comment (lucc) [inline test/functional/legacy/080_substitute_spec.lua]: these where the "headers" in the original test for the different parts. To say it differently: The original test was devided by TEST_X strings and I used these places to split it into seperated it() blocks. I left the strings there to make it easier to incorporate changes from upstream.

**Comment by Pyrohh** [comment]
Looks pretty good so far, although in the future I think having two commits, as you usually do, would be easier to review.

File: test/functional/legacy/080_substitute_spec.lua (hunk 5 above)
Comment (lucc) [inline test/functional/legacy/080_substitute_spec.lua]: like what?

**Comment by lucc** [comment]
I always try to make two (or more commits) but for this test the migration went along very slowly and with errors and problems so I started to modernize it early. Later on I did not manage to split the history in a reasonable manner. I'm sorry I will try to make it better again, next time.

File: test/functional/legacy/080_substitute_spec.lua (hunk 3 above)
Comment (Pyrohh) [inline test/functional/legacy/080_substitute_spec.lua]: That sounds reasonable to me, although maybe it should be as a comment above the `it()`.

File: test/functional/legacy/080_substitute_spec.lua (hunk 5 above)
Comment (Pyrohh) [inline test/functional/legacy/080_substitute_spec.lua]: Heh, I have no idea. It's really not much of an issue.

**Comment by Pyrohh** [comment]
That's fine; the new tests aren't that hard to follow.

**Comment by lucc** [comment]
rebased and updated comments in the file and the commit message. I didn't change the name of the local function.

File: test/functional/legacy/080_substitute_spec.lua [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    eq('o\no', eval([[substitute('oOo', 'O', '\n', '')]]))
+    eq('p\bp', eval([[substitute('pPp', 'P', '\b', '')]]))
+    eq('q\tq', eval([[substitute('qQq', 'Q', '\t', '')]]))
+    eq('r\\r', eval([[substitute('rRr', 'R', '\\', '')]]))
+    eq('scs', eval([[substitute('sSs', 'S', '\c', '')]]))
+    eq('t\rt', eval([[substitute('tTt', 'T', "\r", '')]]))
+    eq('u\nu', eval([[substitute('uUu', 'U', "\n", '')]]))
+    eq('v\bv', eval([[substitute('vVv', 'V', "\b", '')]]))
+    eq('w\\w', eval([[substitute('wWw', 'W', "\\", '')]]))
+    eq('XxxX', eval([[substitute('X', 'X', '\L\uxXx\l\EX', '')]]))
```
Comment (Pyrohh) [inline test/functional/legacy/080_substitute_spec.lua]: This doesn't seem to match the old test:
```
$put =substitute('xXx', 'X', \"\r\", '')
```

File: test/functional/legacy/080_substitute_spec.lua [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    eq('yYYy', eval([[substitute('Y', 'Y', '\U\lYyY\u\Ey', '')]]))
+  end
```
Comment (Pyrohh) [inline test/functional/legacy/080_substitute_spec.lua]: It looks like this case is missing:
```
put =substitute('Z', 'Z', '\U\lZzZ\u\Ez', '')
```

File: test/functional/legacy/080_substitute_spec.lua (hunk 6 above)
Comment (Pyrohh) [inline test/functional/legacy/080_substitute_spec.lua]: Oh, I see. You copied from just test 2, which differs a little bit from test 1:
```diff
--- t1	Fri Jul 10 22:28:16 2015
+++ t2	Fri Jul 10 22:27:55 2015
@@ -1,7 +1,7 @@
 STARTTEST
-:set magic
+:set nomagic
 :set cpo&
-:$put =\"\n\nTEST_1:\"
+:$put =\"\n\nTEST_2:\"
 :$put =substitute('A', 'A', '&&', '')
 :$put =substitute('B', 'B', '\&', '')
 :$put =substitute('C123456789', 'C\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)\(.\)', '\0\9\8\7\6\5\4\3\2\1', '')
@@ -23,11 +23,11 @@
 :$put =substitute('qQq', 'Q', '\t', '')
 :$put =substitute('rRr', 'R', '\\', '')
 :$put =substitute('sSs', 'S', '\c', '')
+:$put =substitute('tTt', 'T', \"\r\", '')
 :$put =substitute('uUu', 'U', \"\n\", '')
 :$put =substitute('vVv', 'V', \"\b\", '')
 :$put =substitute('wWw', 'W', \"\\\", '')
-:$put =substitute('xXx', 'X', \"\r\", '')
-:$put =substitute('Y', 'Y', '\L\uyYy\l\EY', '')
-:$put =substitute('Z', 'Z', '\U\lZzZ\u\Ez', '')
-/^TEST_2
+:$put =substitute('X', 'X', '\L\uxXx\l\EX', '')
+:$put =substitute('Y', 'Y', '\U\lYyY\u\Ey', '')
+/^TEST_3
 ENDTEST

```

**Comment by Pyrohh** [comment]
This looks pretty good to me, but I'll try to review it once more later. Please feel free to ping me if I forget.

**Comment by lucc** [comment]
@Pyrohh latest inline comments: I'm trying to explain this in the commit message. The tests are identical in the sense that they test the same patterns. They only use different surrounding letters. If you look at the original lines for t, x, y, z you will notice what I mean. Maybe I should add a comment to the top of the function again:
```lua
-- This function was abstracted from the original part one and two of the legacy tests. Tese parts where functional identical as they tested the same patterns. The only difference where the surrounding letters of some test lines (namely the `t`, `x`, `y` and `z` lines). These lines have been reordered here.
```

File: test/functional/legacy/080_substitute_spec.lua [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+      [[submatch(7, 1), submatch(6, 1), submatch(5, 1), submatch(4, 1), ]] ..
+      [[submatch(3, 1), submatch(2, 1), submatch(1, 1)])', '')]]))
+  end)
+
+  it('with "set cpo±=/" (TEST_6)', function()
+    execute('set magic&')
+    execute('set cpo+=/')
+    eq('a', eval("substitute('A', 'A', 'a', '')"))
+    eq('%', eval("substitute('B', 'B', '%', '')"))
+    execute('set cpo-=/')
```
Comment (Pyrohh) [inline test/functional/legacy/080_substitute_spec.lua]: Now that #2943 was merged, line 116 up to and including 119 should be removed.

**Comment by Pyrohh** [comment]
@lucc Could you please address my comment and rebase to fix the merge conflict?

**Comment by Pyrohh** [comment]
Also, I'm not sure the `set &cpo` calls are needed anymore.

File: test/functional/legacy/080_substitute_spec.lua (hunk 8 above)
Comment (lucc) [inline test/functional/legacy/080_substitute_spec.lua]: done

**Comment by lucc** [comment]
I removed the cpo& stuff but kept the magic& for clarity.

**Comment by Pyrohh** [comment]
LGTM, merged. Thanks a lot @lucc.

**Comment by lucc** [comment]
thanks @Pyrohh test79 is very similar, if you want to have a look, it is in #2988.

**Comment by Pyrohh** [comment]
sure, I'll take a look at that.
